### PR TITLE
Add HttpListener tests and fix a porting bug

### LINKED ISSUE #17980: NullReferenceException thrown getting InputStream from HttpListener
```cs
class Program
{
    static async Task Test()
    {
        string uri = "http://localhost:9200/";
        HttpListener listener = new HttpListener();
        listener.Prefixes.Add(uri);
        listener.Start();

        using (ClientWebSocket client = new ClientWebSocket())
        {
            UriBuilder uriBuilder = new UriBuilder(uri) { Scheme = "ws" };
            Task<HttpListenerContext> serverContextTask = listener.GetContextAsync();

            Task clientConnectTask = client.ConnectAsync(uriBuilder.Uri, CancellationToken.None);

            HttpListenerContext listenerContext = await serverContextTask;
            HttpListenerWebSocketContext wsContext = await listenerContext.AcceptWebSocketAsync(null);
            await clientConnectTask;

            HttpListenerRequest request = listenerContext.Request;
            Stream stream = request.InputStream;
        }

        listener.Stop();
    }

    static void Main()
    {
        Test().Wait();
    }
}
```
Stack trace:
```
     System.NullReferenceException : Object reference not set to an instance of an object.
      Stack Trace:
            at System.Net.HttpListenerRequest.get_ContentLength64()
            at System.Net.HttpListenerRequest.get_HasEntityBody()
            at System.Net.HttpListenerRequest.get_InputStream()
```

I think this is a pretty serious bug - it makes reading any input from HttpListener's tream broken. Netfx works, obviously.

I've got a fix for this, and adding tests (as per #13618)

The reference source is:
```
 if (HttpWebRequest.ChunkedHeader.Equals(GetKnownHeader(HttpRequestHeader.TransferEncoding)
```

However, the .NET core is
```
 if (GetKnownHeader(HttpRequestHeader.TransferEncoding).Equals(HttpWebRequest.ChunkedHeader))
```

This is *not* the same!

### PR DESCRIPTION
**Pull request by hughbe**
This covers `HttpListenerRequestStream` and `HttpListenerResponseStream`.

The tests will fail with NulLReferenceExceptions without Commit 2, which fixes #17980, a bug caused porting HttpListener to corefx.

@Priya91 @davidsh @CIPop

### CODE REVIEW

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                client.DefaultRequestHeaders.TransferEncodingChunked = transferEncodingChunked;
+                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent(Text));
+
+                HttpListenerContext context = await contextTask;
+
+                // Disconnect the client from the HttpListener.
+                client.Dispose();
+
+                // TODO: we need to sleep and wait to make sure the client closes.
+                Thread.Sleep(1000);
```
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: I'm not sure the best way to fix this TODO. I'm trying to test a scenario where we have connected to the client/socket, but fail to read from or write to it. The same goes for the other TODOs.

Without the thread sleep, these tests fail intermittently.

Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: This test requires knowing exactly when the connection has been closed, which is abstracted away in HttpClient.  It'd probably be best to just use a Socket as the client directly, e.g.
```C#
using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
{
    client.Connect(_factory.ListeningUrl.Host, 80);
    client.Send(Encoding.ASCII.GetBytes($"POST {_factory.ListeningUrl.PathAndQuery} HTTP/1.1\r\nHost: {_factory.ListeningUrl.Host}\r\nContent-Length: {Text.Length}\r\n\r\n{Text}"));
}
```
etc. just as a rough sketch.

File: src/System.Net.HttpListener/src/System/Net/Windows/HttpListenerRequest.Windows.cs [hunk 2]
```diff
@@ -232,7 +232,8 @@ public long ContentLength64
             {
                 if (_boundaryType == BoundaryType.None)
                 {
-                    if (GetKnownHeader(HttpRequestHeader.TransferEncoding).Equals("chunked", StringComparison.OrdinalIgnoreCase))
+                    string transferEncodingHeader = GetKnownHeader(HttpRequestHeader.TransferEncoding);
+                    if (transferEncodingHeader != null && transferEncodingHeader.Equals("chunked", StringComparison.OrdinalIgnoreCase))
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/HttpListenerRequest.Windows.cs]: Good fix, thanks.

File: src/System.Net.HttpListener/tests/Common.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        public GetContextHelper(HttpListener listener, string listeningUrl)
+        {
+            _listener = listener;
+            _listeningUrl = listeningUrl;
+            _client = new HttpClient();
+        }
+
+        public async Task<HttpListenerResponse> GetResponse()
+        {
+            // We need this to avoid hanging tests.
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/Common.cs]: Why? Can you explain further?

File: src/System.Net.HttpListener/tests/Common.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            Task<string> clientTask = _client.GetStringAsync(_listeningUrl);
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/Common.cs]: We're never waiting on clientTask.  What if it fails?

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+    {
+        private HttpListenerFactory _factory;
+        private HttpListener _listener;
+        private GetContextHelper _helper;
+
+        public HttpRequestStreamTests()
+        {
+            _factory = new HttpListenerFactory();
+            _listener = _factory.GetListener();
+            if (!PlatformDetection.IsWindows7)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: Why is this needed?

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: Thanks for writing a bunch of tests!

I'm curious; did you look at Mono to see what tests it has?  If it has good ones, would you want to look at porting them in addition to these?

File: src/System.Net.HttpListener/tests/Common.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/Common.cs]: I'm not too sure why I needed this if I'm honest. Bascially the code `await listener.GetContextAsync()` never ends if we don't have something trying to connect to the listener. In this case, `_client.GetStringAsync(_listeningUrl)` prevents `await listener.GetContextAsync()` from never exiting - why, I'm not sure.

Comment (stephentoub) [inline src/System.Net.HttpListener/tests/Common.cs]: Oh, it's in reference to using the client?  That's necessary because GetContextAsync is about handling a request, so if you never get a request, it'll obviously never complete.

Comment (hughbe) [inline src/System.Net.HttpListener/tests/Common.cs]: As I thought. I added a comment in the PR feedback commit

File: src/System.Net.HttpListener/tests/Common.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/Common.cs]: See comment above in updated PR feedback commit

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: ClientWebSocket isn't supported in Windows 7. I have updated the PR and this whole code block and reliance on ClientWebSocket has gone.

**Comment by hughbe** [comment]
Okay, looks like a couple of differences between Windows and Unix. Sorry, I only have a Windows box. What I'll do is I'll disable these tests on Unix in the next round of PR feedback and submit another PR fixing these tests if appropriate. This PR is already large enough (sorry!)

File: src/System.Net.HttpListener/tests/Common.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    public class CustomAsyncResult : IAsyncResult
+    {
+        public object AsyncState => throw new NotImplementedException();
+        public WaitHandle AsyncWaitHandle => throw new NotImplementedException();
+        public bool CompletedSynchronously => throw new NotImplementedException();
+        public bool IsCompleted => throw new NotImplementedException();
+    }
+
+    public class Common
+    {
```
Comment (Priya91) [inline src/System.Net.HttpListener/tests/Common.cs]: This is used only for asynccallback in the stream read operations. Why not use null for callback and remove this stub?

File: src/System.Net.HttpListener/tests/System.Net.HttpListener.Tests.csproj [hunk 8]
```diff
@@ -11,6 +11,9 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netstandard-Release|AnyCPU'" />
   <ItemGroup>
     <Compile Include="AuthenticationTests.cs" />
+    <Compile Include="Common.cs" />
```
Comment (Priya91) [inline src/System.Net.HttpListener/tests/System.Net.HttpListener.Tests.csproj]: Rename this file to GetContextHelper, since that's the major class in the file.

**Comment by Priya91** [comment]
@hughbe These tests fail on windows debug build as well

```
System.Net.Tests.HttpResponseStreamTests.Write_HeadersToClosedConnectionSynchronously_ThrowsHttpListenerException(ignoreWriteExceptions: True)
System.Net.Tests.HttpResponseStreamTests.Write_ContentToClosedConnectionSynchronously_ThrowsHttpListenerException(ignoreWriteExceptions: False)
```
Link: https://ci.dot.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/5841/

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+
+            using (HttpClient client = new HttpClient())
+            {
+                client.DefaultRequestHeaders.TransferEncodingChunked = transferEncodingChunked;
+                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent(text));
+
+                HttpListenerContext context = await contextTask;
+
+                // If the size is greater than 128K, then we limit the size, and have to do multiple reads.
+                byte[] buffer = new byte[expected.Length];
```
Comment (Priya91) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: This limit is coming from http.sys, windows native stack that httplistener uses. We don't have this restriction in managed implementation. This will fail on Unix.

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
```
Comment (Priya91) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: Same as before..

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+                IAsyncResult beginReadResult = inputStream1.BeginRead(new byte[0], 0, 0, new AsyncCallback(Common.StubCallback), null);
+
+                Assert.Throws<ArgumentException>("asyncResult", () => inputStream2.EndRead(new CustomAsyncResult()));
+                Assert.Throws<ArgumentException>("asyncResult", () => inputStream2.EndRead(beginReadResult));
+            }
+        }
+
+        [Fact]
+        public async Task EndRead_CalledTwice_ThrowsInvalidOperationException()
+        {
```
Comment (Priya91) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: This is failing on Unix, due to difference in implemenation. The async result on Unix, doesn't use lazyasyncresult, and we don't check if end was already called. This behavior is not consistent across all Streams. For ex, Network stream doesn't throw for multiple endread calls. cc @stephentoub Should we mimic windows behavior on Unix for this one?

**Comment by Priya91** [comment]
@hughbe Most the test failues on request stream and response stream on Unix are due to differences in input parameter invalid conditions checking. We should unify these checks where possible. You can either skip these tests and file bug, or fix it as well. Thanks for working on this :)

**Comment by danmosemsft** [comment]
>  I only have a Windows box.

@hughbe many of us use a Unix VM on our Windows box rather than having a dedicated Unix box. Either with HyperV or VirtualBox, both of which are free. No reason you should set this up, but just FYI in case it wasn't obvious.

**Comment by hughbe** [comment]
Thanks for the review @priya91 ill address your comments soon.

Quick question for you or @stephentoub: I'm really not sure why the writing to a closed socket test are failing on Windows. They pass on my machine every time (in the latest commit I ran them 15 times). Obviously what's happening is the socket isn't closing properly. Do you know the best way to ensure the socket is fully closed before writing to it - the GC collection and waiting seems to do the trick for me but not the CI. 

Since the pr is already quite long I'll file an issue for fixing the unix implementation, with the intent to grab that issue and fix the differences myself. I'll mark the failing tests as activeissue for anyunix.

Thanks Dan, I keep saying I'll get round to it but this should be a good motivation!

**Comment by stephentoub** [comment]
> I'm really not sure why the writing to a closed socket test are failing on Windows

Wich code are you referring to?

**Comment by hughbe** [comment]
Write_HeadersToClosedConnectionSynchronous_ThrowsHttpListenerException and Write_ContentToClosedConnectionSynchronous_ThrowsHttpListenerException in https://github.com/hughbe/corefx/blob/67fa78b295fbb245daf47a18322913d426e4089c/src/System.Net.HttpListener/tests/HttpResponseStreamTests.cs

File: src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs (hunk 6 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs]: I agree - it's a good idea to port over the tests from Mono. I've taken a look at the Mono tests - not many duplicates with those added in this PR. The sheer number of Unix failures on this PR bolster this reasoning.

**Comment by Priya91** [comment]
@hughbe Socket.dispose, is calling safehandle.dispose. It could be the case that there were some references to the safe handle that prevents it from disposing, and with explicilty waiting for gc, that got disposed. Or it could be that the safehandle was put in finalizer queue, and the finalizer hadn't run to free up the resources before you tried to use the stream on the native connection.
